### ISSUE TITLE
Travis manifest_upload job occasionally fails to create merge_pr_* tags or upload manifest

### ISSUE THREAD

**Issue by foolip** [web-platform-tests/wpt#10572]
Running a variation of the script in https://github.com/foolip/ecosystem-infra-stats/issues/6#issuecomment-375731858 to make sure that all PRs merged after 2017-07-01 have been tagged, in particular ones created well before that point, I noticed that some recently merged PRs have also not been tagged.

https://github.com/w3c/web-platform-tests/pull/10543 is the most recent. The PR and GitHub API says the PR was merged in commit 0628e2e, but that's #10124. There are two Travis jobs appear to have been run with that as checked out revision: https://travis-ci.org/w3c/web-platform-tests/jobs/369319256 and https://travis-ci.org/w3c/web-platform-tests/jobs/369365460. The second says "ERROR:\_\_main\_\_:Tag creation failed", but the first does not. At the very least something seems to have gone wrong on GitHub's side here.

https://github.com/w3c/web-platform-tests/pull/10445 was processed in https://travis-ci.org/w3c/web-platform-tests/jobs/365663222 but log says "ERROR:\_\_main\_\_:No PR found for master". Why?

For https://github.com/w3c/web-platform-tests/pull/10194 it seems like no build was triggered, in https://travis-ci.org/w3c/web-platform-tests/builds I see only the one before and after: https://travis-ci.org/w3c/web-platform-tests/builds/361106934 + https://travis-ci.org/w3c/web-platform-tests/builds/361263818

Those three are the only that are not tagged since #8005 was merged, the rest were merged before that, the latest of those being https://github.com/w3c/web-platform-tests/pull/4974.

**Comment by csnardi** [comment]
The changes made in #10543 were already made in #10502 -- I assume GitHub when rebasing & merging created an empty commit and so that was swallowed up somehow (not sure how git actually deals with that, but it would make sense).

**Comment by foolip** [comment]
Oh, that would explain it indeed. @jgraham, in this situation, it seems like rather than treat the PR as merged, it would be better to close it? In any case it's appropriate that no merge_pr_* tag was created, because nothing on master actually happened, or at least commit 0628e2e73489ddd075f273d2555e504e97e28ce6 is no better a place to put the tag than some of the previous commits.

**Comment by foolip** [comment]
I have created the merge_pr_10194 and merge_pr_10445 tags, and deleted the untagged-9aa6616bd18782ef19ad that was automatically created.

https://github.com/w3c/web-platform-tests/pull/10543 is still an odd case that will need @jgraham's input.

**Comment by foolip** [comment]
@jgraham the problem is that the PR claims to be merged at commit 0628e2e73489ddd075f273d2555e504e97e28ce6, but that's just what master happened to be when the rebase happened. And https://github.com/w3c/web-platform-tests/pull/10543/files still lists changes that weren't made as part of that PR.

Seems to me that the rebase must have resulted in no commits on the branch (not an empty commit), and that case could instead be handled by abandoning the PR, that would be less confusing to understand after the fact.

**Comment by foolip** [comment]
It just happened again. untagged-f360b6391e175d60f813 was created for commit 4f397167b4ed552a02201c92d363cfaecfe2c7f0 which is from https://github.com/w3c/web-platform-tests/pull/10663, and the build was https://travis-ci.org/w3c/web-platform-tests/builds/371818852.

This time the log says:
```
++python tools/ci/tag_master.py
ERROR:__main__:Tag creation failed:
HTTP Error 422: Unprocessable Entity
```

Don't know why that is, maybe a GitHub API problem.

**Comment by foolip** [comment]
https://github.com/w3c/web-platform-tests/pull/10673 is to make the Travis job fail if no merge_pr_* tag could be created, but it won't help with the problem just reported in https://github.com/w3c/web-platform-tests/issues/10572#issuecomment-384906224.

**Comment by foolip** [comment]
I restarted https://travis-ci.org/w3c/web-platform-tests/jobs/371818853 but that didn't fix the problem, now it says "ERROR:\_\_main\_\_:No PR found for master" instead.

**Comment by foolip** [comment]
I have deleted untagged-f360b6391e175d60f813 and created merge_pr_10663 instead.

**Comment by foolip** [comment]
https://github.com/w3c/web-platform-tests/pull/10673 worked, because we just had a broken build from master: https://travis-ci.org/w3c/web-platform-tests/jobs/372015780, which is for https://github.com/w3c/web-platform-tests/pull/10678.

That PR doesn't show up in https://api.github.com/search/issues?q=type:pr+is:merged+repo:w3c/web-platform-tests+438f2694a95cb9c7f329ef89c1d37fa7140106f0 even now, so restarting it can't help. @annevk, I presume you did nothing unusual when merging that, just used the buttons?

**Comment by foolip** [comment]
The PR merged immediately before it is https://github.com/w3c/web-platform-tests/pull/10675 and https://api.github.com/search/issues?q=type:pr+is:merged+repo:w3c/web-platform-tests+cc29a15b5c00d90149d8f60f96ca4f77a989a2d7 does find that, but I can't see what the difference could be.

The temptation to blame everything odd on GitHub is strong, should we call this "search indexing is broken"? dunno

**Comment by annevk** [comment]
Right, per https://github.com/w3c/web-platform-tests/commits/master it's been broken for a while.

**Comment by foolip** [comment]
Note that `travis_retry` probably won't fix this problem unless it also has waiting built in. Otherwise it'll just try n times fast and fail anyway.

**Comment by foolip** [comment]
OK, job restart worked and created merge_pr_10678 + merge_pr_10671, I didn't do those manually.

**Comment by foolip** [comment]
The previous and probably continued unreliability of this came up in https://github.com/web-platform-tests/wpt.fyi/pull/101, where @lukebjerring's adding a wpt.fyi API endpoint to get the manifest, backed by the GitHub tags/releases. https://github.com/web-platform-tests/wpt.fyi/pull/101#issuecomment-386246656 has a list of tags for which there's no manifest.

@lukebjerring, elsewhere (can't find it) you had a suggestion for making the tagging more reliable by inspecting all commits since the last tagged commit. Something similar for manifest upload might make sense, where we first try to find the oldest tag without an uploaded manifest and move forward from there.

Both of those will be a bit odd to trigger by changes to master, but maybe that doesn't matter.

**Comment by lukebjerring** [comment]
Yes, that was my suggestion (try all commits since the last successful tag). 

It is a bit weird (write-on-read kind of vibe) to have master trigger those changes, but unless we want an arbitrarily timed cron job, commits to master still makes the most sense (they're quite regular).

Note that the change would still sometimes succeed the race condition anyway, so you'd see equivalent to current behaviour in the good cases, and a fallback in others.

**Comment by foolip** [comment]
I think tagging retroactively is fine. For generating the manifest we'd need to check out earlier revisions, and then revert back to the commit originally checked so that later Travis steps aren't confused, so that introduced some small risk of breaking things in novel ways.

@jgraham, WDYT about this approach for tags and/or manifest upload? The only other option is something cron-like, perhaps a daily Travis cron job, but that'd be another thing to monitor for errors...

**Comment by mdittmer** [comment]
Ping @foolip. This is labeled priority:roadmap but has no owner. Does it align with any infra folks' core objectives this quarter?

**Comment by foolip** [comment]
It has happened once in the past month or so that a untagged-* tag was created, which I then deleted. In https://github.com/web-platform-tests/wpt/releases a lot of "Draft" releases with no assets are listed, and some with assets. Why this happens I don't know, but I'll delete them. I'm also writing a script to check if recent merged PRs have tags and releases with assets.

**Comment by foolip** [comment]
Wrote this script, but it hasn't finished running :) (needs nodejs 10)
```js
const octokit = require('@octokit/rest')()

const repoOptions = { owner: 'web-platform-tests', repo: 'wpt' };

// merge_pr_* tags should exist since July 2017.
const SINCE = Date.parse('2017-07-01T00:00Z');

// gets all PRs with pagination
async function* getAllPullRequests(prOptions) {
    let page = 1;
    while (true) {
        const prs = (await octokit.pullRequests.getAll({
            ...repoOptions,
            ...prOptions,
            per_page: 100,
            page,
        })).data;
        if (prs.length === 0) {
            break;
        }
        for (const pr of prs) {
            yield pr;
        }
        page++;
    }
}

async function main() {
    octokit.authenticate({
        type: 'token',
        token: process.env.GH_TOKEN,
    });

    // Note: sorting by update time means the order can change, so with
    // pagination we may miss some PRs and see some twice. But the
    // alternative is to fetch every single PR, since old triple-digit
    // PRs can be merged at any time and should be tagged.
    const prs = getAllPullRequests({
        state: 'closed',
        sort: 'updated',
        direction: 'desc',
    });

    for await (const pr of prs) {
        if (Date.parse(pr.updated_at) < SINCE) {
            console.log(`Stopping search at PR #${pr.number} updated at ${pr.updated_at}`)
            break;
        }

        // Skip unmerged and old PRs
        if (!pr.merged_at || Date.parse(pr.merged_at) < SINCE) {
            continue;
        }

        const tag = `merge_pr_${pr.number}`;

        let release;
        try {
            release = (await octokit.repos.getReleaseByTag({
                ...repoOptions,
                tag
            })).data;
        } catch (e) {
            // no release, check if there's a tag
            try {
                await octokit.gitdata.getReference({
                    ...repoOptions,
                    ref: `tags/${tag}`
                });
                // there is a tag, just no release
                console.log(`${tag}: no release`);
            } catch(e) {
                console.log(`${tag}: no tag`);
            }
            continue;
        }

        const manifest = release.assets.find(asset => {
            return /^MANIFEST-[0-9a-f]{40}.json.gz$/.test(asset.name);
        });

        if (!manifest) {
            console.log(`${tag}: no manifest`);
            continue;
        }

        // The (compressed) manifest should be >2MB.
        if (manifest.size < 2000000) {
            console.log(`${tag}: manifest too small (${manifest.size})`);
        }

        console.log(`${tag}: OK`);
    }
}

main();
```

**Comment by foolip** [comment]
However, it's run long enough that I can tell that the latest trouble was just 15 days ago with https://github.com/web-platform-tests/wpt/pull/13187. That was tagged but there's no release, because Travis failed due to `Octokit::BadGateway`:
https://travis-ci.org/web-platform-tests/wpt/jobs/432494382

There are more like that.

The most recent with no tags is https://github.com/web-platform-tests/wpt/pull/12824, where the Travis build associated with commit d7627ae069c4b90d06f9f6ac55b3658849f7e92e (https://travis-ci.org/web-platform-tests/wpt/jobs/432494382) oddly thinks itself to be for commit 00c1fd9ea4bc1e03f966919f1e209ab8cd81f57e and a PR build. This is weird, possibly the result of GitHub or Travis backend bugs.

https://github.com/web-platform-tests/wpt/pull/12854 also got no tag, but for network error reasons:
https://travis-ci.org/web-platform-tests/wpt/jobs/424950964

As far as my script has reached, there are problems with ~1.5% (36/2214) of merged PRs. That's too much, so I'll leave this open.

I'll also manually add some missing tags.

**Comment by foolip** [comment]
Ah, wait, I've made a mistake! https://github.com/web-platform-tests/wpt/pull/12824 wasn't targeting the master branch, so that explains the weirdness. My script should skip PRs that target anything but master.

**Comment by foolip** [comment]
https://github.com/web-platform-tests/wpt/pull/11452 is like https://github.com/web-platform-tests/wpt/issues/10572#issuecomment-383751931 so that case still happens.

**Comment by foolip** [comment]
#10605  had no manifest, again because of a network error in the deploy step. It was recent, so I restarted https://travis-ci.org/web-platform-tests/wpt/jobs/440405376, but that then fails with "HTTP Error 422: Unprocessable Entity" because the tag already exists...

**Comment by foolip** [comment]
Note that there are some tags for which releases exist but the assets (manifest file) do not. Those are merge_pr_10605 and merge_pr_10247, and I'll just delete the releases to put them in the same category as 'no release'.

**Comment by Hexcles** [comment]
Briefly discussed this with Philip today. We need some better deploy mechanisms. We can keep pushing the limits of the CI system, by adding more retries to each step, and making sure later steps only happen when earlier steps succeed (e.g. make sure we have "set -e"). Or rather, it'd be good if there's a better system to define such workflow or pipeline. GitHub just announced [Actions](https://github.com/features/actions) which looks interesting and could be simpler than square pegging the workflow into Travis, but I haven't looked into the details yet.

**Comment by foolip** [comment]
There was a problem today again: https://github.com/web-platform-tests/wpt/pull/13638

**Comment by foolip** [comment]
A novel type of failure in https://travis-ci.org/web-platform-tests/wpt/jobs/454430391, where the deploy step failed with "No output has been received in the last 10m0s".

I wonder if anyone is actually using Travis to deploy stuff at any significant scale :/

As @Hexcles says, this seems like a thing we should eventually migrate to the new Actions.

**Comment by foolip** [comment]
I clicked some buttons as soon as Actions was announced, and the enrollment page now says "This account is already on the waitlist for the GitHub Actions Beta!"

**Comment by Hexcles** [comment]
@foolip that's new for this particular task, but not really novel for Travis in general. Travis often has large-scale task timeouts like this, which seems like transient infra issues. In fact, a bunch of stability checks failed yesterday for the same reason.

**Comment by foolip** [comment]
Yep, until this is fixed we won't be able to depend on having a manifest for wpt.fyi, which would be useful. @lukebjerring do you want to own this, or what's the priority of having the manifest available reliably?

**Comment by lukebjerring** [comment]
No, I don't want to own this.
wpt.fyi falls back to the `latest` manifest when it 404's for a particular SHA, so it's not blocked on having this be reliable, it's just a lower-quality feature (since there can be manifest mismatch in some cases).

**Comment by foolip** [comment]
@lukebjerring glad to hear it's not a blocker for wpt.fyi :)

This is continuing to fail intermittently on Travis CI, https://github.com/web-platform-tests/wpt/runs/50028082 being a recent case.

I've played around with GitHub Actions enough to be confident we can do this using GitHub actions when it exits beta, but until then let's leave it alone.

**Comment by zcorpan** [comment]
https://travis-ci.com/web-platform-tests/wpt/jobs/177122640 might be another case of this issue.
```
++python tools/ci/tag_master.py
ERROR:__main__:Tag creation failed:
HTTP Error 422: Unprocessable Entity
```
I restarted the job but it errored again. I don't know how to access the log of the first attempt. The [PR](https://github.com/web-platform-tests/wpt/pull/15357) was green.

**Comment by foolip** [comment]
After a restart the logs of previous runs are gone I think.

I've seen Error 422 before when working on this, but I don't remember when it happens.

**Comment by foolip** [comment]
This was moved to GitHub Actions in https://github.com/web-platform-tests/wpt/pull/15781, so I'll close this.

GitHub Actions is probably not 100% reliable either, but I'll file another issue when it becomes a problem.
